### PR TITLE
Change the way tensor is reshaped in BartAttention (from .view to .reshape)

### LINKED ISSUE #21813: Error when using BART for Prefix Tuning. Replace `view` with `reshape` in `BartAttention`?
### System Info

- `transformers` version: 4.26.1
- Platform: Linux-5.10.147+-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.13.1+cu116 (False)
- Tensorflow version (GPU?): 2.11.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker @younesbelkada

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

When using Prefix Tuning with BART in PEFT, an error occurs for some edge cases, see [#129](https://github.com/huggingface/peft/issues/129#issue-1598538584), with a suggestion to replace `view` with `reshape` in `BartAttention`.

### Expected behavior

I would expect the example code provided in [#129](https://github.com/huggingface/peft/issues/129#issue-1598538584) to work regardless of the length of the input.

### PR DESCRIPTION
**Pull request by raghavanone**
# What does this PR do?

Fixes #21813

### CODE REVIEW

**Comment by raghavanone** [comment]
@younesbelkada For some reason fix-copies is not fixing the prophetnet copy, Not sure how to fix this.

**Comment by younesbelkada** [comment]
Hi @raghavanone !
Hum very weird, can you try to merge this branch with the `main` branch of `transformers` and see if this fixes the issue?

**Comment by raghavanone** [comment]
> Hi @raghavanone ! Hum very weird, can you try to merge this branch with the `main` branch of `transformers` and see if this fixes the issue?

Indeed weird, It is already on top of the latest main. Stranger thing is both of these checks pass on my machine.

**Comment by younesbelkada** [comment]
Can you try `pip install --upgrade -e .["quality"]` + `make fixup`  + `make fix-copies` ?

File: tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py [hunk 1]
```diff
@@ -152,7 +152,10 @@ def test_model(self):
         self.model_tester.create_and_check_model(*config_and_inputs)
 
     @unittest.skip(
-        reason="skip for now as the computed `max_memory` by `model_split_percents` in the test method will be changed inside `from_pretrained`"
+        reason=(
```
Comment (younesbelkada) [inline tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py:155]: I think all the changes inside `tests` shouldn't be there, can you please revert them !

File: src/transformers/models/efficientnet/convert_efficientnet_to_pytorch.py [hunk 2]
```diff
@@ -324,7 +324,9 @@ def convert_efficientnet_checkpoint(model_name, pytorch_dump_folder_path, save_m
         "--model_name",
         default="b0",
         type=str,
-        help="Version name of the EfficientNet model you want to convert, select from [b0, b1, b2, b3, b4, b5, b6, b7].",
+        help=(
+            "Version name of the EfficientNet model you want to convert, select from [b0, b1, b2, b3, b4, b5, b6, b7]."
+        ),
```
Comment (younesbelkada) [inline src/transformers/models/efficientnet/convert_efficientnet_to_pytorch.py:329]: this change should not be there

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this awesome work! 
I am ok with this PR in the principle that we are calling `reshape` on the same tensor (i.e. we are not creating a new tensor on top of `reshape`) and this PR might un-lock the issues observed with some supported libraries such as `peft`.
 
My understanding is that `view` will always share data with the original tensor, which might be problematic in the case we create a new tensor
```python
x = z.view(*shape)
z.fill_(1)
print(x)
>>> tensor([1., 1., 1.])
```
Also, according to the `reshape` [documentation](https://pytorch.org/docs/master/generated/torch.reshape.html#torch.reshape):

> Returns a tensor with the same data and number of elements as input, but with the specified shape. When possible, the returned tensor will be a view of input. Otherwise, it will be a copy. Contiguous inputs and inputs with compatible strides can be reshaped without copying, but you should not depend on the copying vs. viewing behavior.

Since the modeling files worked fine with `view`, I expect `reshape` to return a view of the tensor and not a copy

I would love to hear from @ArthurZucker & @sgugger to see what they say

File: src/transformers/models/efficientnet/convert_efficientnet_to_pytorch.py (hunk 2 above)
Comment (raghavanone) [reply src/transformers/models/efficientnet/convert_efficientnet_to_pytorch.py:329]: Done

File: tests/test_pipeline_mixin.py [hunk 3]
```diff
@@ -217,24 +217,24 @@ def run_pipeline_test(self, task, repo_name, model_architecture, tokenizer_name,
...
-        # TODO: We should check if a model file is on the Hub repo. instead.
+        # TODO: We should check if a model file is on the Hub repo instead.
         try:
             model = model_architecture.from_pretrained(repo_id)
         except Exception:
             self.skipTest(
-                f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')} is skipped: Could not find or load "
-                f"the model from `{repo_id}` with `{model_architecture}`."
+                f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')} is skipped: Could not find or load"
+                f" the model from `{repo_id}` with `{model_architecture}`."
```
Comment (younesbelkada) [inline tests/test_pipeline_mixin.py:237]: Please also revert these changes

File: tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py (hunk 1 above)
Comment (raghavanone) [reply tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py]: Done

File: tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py [hunk 4]
```diff
@@ -182,7 +182,10 @@ def test_model(self):
...
+            "skip for now as the computed `max_memory` by `model_split_percents` in the test method will be changed"
+            " inside `from_pretrained`"
+        )
```
Comment (younesbelkada) [inline tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py:188]: Here as well

File: tests/test_pipeline_mixin.py (hunk 3 above)
Comment (raghavanone) [reply tests/test_pipeline_mixin.py]: Done

File: tests/test_pipeline_mixin.py [hunk 5]
```diff
@@ -224,11 +224,11 @@ def run_pipeline_test(self, task, repo_name, model_architecture, tokenizer_name,
...
         if tokenizer is None and processor is None:
             self.skipTest(
-                f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')} is skipped: Could not find or load "
-                f"any tokenizer / processor from `{repo_id}`."
+                f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')} is skipped: Could not find or load"
+                f" any tokenizer / processor from `{repo_id}`."
             )
 
-        # TODO: We should check if a model file is on the Hub repo. instead.
+        # TODO: We should check if a model file is on the Hub repo instead.
```
Comment (ArthurZucker) [inline tests/test_pipeline_mixin.py:231]: This should probably not be in this PR as it is unrelated!

Comment (younesbelkada) [inline tests/test_pipeline_mixin.py:231]: This should be reverted as well

**Review by ArthurZucker — APPROVED** [review]
LGTM, let's make sure that the tests pass, but otherwise thanks for the fix. (also let's remove unrelated changes

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for this!

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!
